### PR TITLE
Implement fill_random with execution space

### PR DESCRIPTION
**Pull request by masterleinad**
@kyungjoo-kim requested this on slack. Part of https://github.com/kokkos/kokkos/issues/4792.

### CODE REVIEW

**Comment by kyungjoo-kim** [comment]
thank you.

**Review by PhilMiller — APPROVED** [review]

**Comment by masterleinad** [comment]
`OpenMPTarget` failing in `openmptarget.unique_token_instance` looks spurious:
```
 RUN      ] openmptarget.unique_token_instance
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestUniqueToken.hpp:151: Failure
4: Expected equality of these values:
4:   sum
4:     Which is: 9999990
4:   int64_t(N) * R
4:     Which is: 10000000
4: [  FAILED  ] openmptarget.unique_token_instance (229 ms)
```

**Review by fnrizzi — APPROVED** [review]

File: algorithms/unit_tests/TestRandom.hpp [hunk 1]
```diff
@@ -522,13 +522,15 @@ struct TestDynRankView {
...
 
     ReducerValueType val;
-    Kokkos::parallel_reduce(Kokkos::RangePolicy<ExecutionSpace>(0, A.size()),
-                            *this, ReducerType(val));
+    Kokkos::parallel_reduce(
+        Kokkos::RangePolicy<ExecutionSpace>(exec, 0, A.size()), *this,
+        ReducerType(val));
 
-    Kokkos::fence();
+    exec.fence();
```
Comment (dalg24) [inline algorithms/unit_tests/TestRandom.hpp:533]: We don't even need a fence here do we?

Comment (PhilMiller) [reply algorithms/unit_tests/TestRandom.hpp:533]: Is that the semantic for this particular overload of `parallel_reduce`?

Regardless, does it hurt anything for the test to have it?

**Pull request merged**